### ISSUE TITLE
AGS 4: Make linking events to functions more intuitive

### ISSUE THREAD

**Issue by messengerbag** [adventuregamestudio/ags#2550]
*Problem**
For many years, one of the most common support requests from new AGS users is that they can't get an event handler to run. This is almost always because they've forgotten (or don't realize that they need) to link it to the corresponding event via the event pane. Even experienced users sometimes forget this step, adding delays as they test the game in order to discover and fix the oversight.

There is also an inconsistency in the fact that [global event handlers](https://adventuregamestudio.github.io/ags-manual/Globalfunctions_Event.html) don't need to be linked, but use a predefined name that is called at the right time if it exists.

**Suggested change**
See [previous forum discussion](https://www.adventuregamestudio.co.uk/forums/editor-development/feature-suggestion-auto-link-event-handlers/), where several alternative solutions were proposed, including:

- Detecting unlinked functions that match the default event handler names at build time and displaying a warning
- Automatically link event handlers by default, without a need to do so manually
- Marking linked event handlers in the script editor (in the left-hand column), and a warning icon if not linked 
- Link event handlers explicitly in script rather than through the event pane, using delegates

Perhaps there are new ideas now. One possible "quick fix" is to simply improve the documentation of this feature/quirk, for example in an FAQ section of the manual.

**Comment by ivan-mogilko** [comment]
For the reference, note that after #2534 is merged it is possible to have the handler functions in any module (except for room and room object events). The module is selected by the special property per global object (gui controls share one with their parent GUI).

**Comment by ericoporto** [comment]
The reverse issue is documented in the ticket https://github.com/adventuregamestudio/ags/issues/1363

**Comment by ivan-mogilko** [comment]
I have a uneasy feeling about automatic linking, because if we implement delegates then linking any event with a function  without clear instruction may be contradicting to user's intent.


I'd like to suggest another option: introduce annotations in script, which may be added before the function and instruct editor to link them. This is not necessarily a only solution, but maybe some will like to have.

**Comment by AlanDrake** [comment]
I've always been told that annotations are evil.
Though you are right to be uneasy, the source of uneasiness in the first place was that users had to manually associate functions, which likely led them to create the wrong type of event and later relink it elsewhere.

Were there no need to link manually from the editor, a user could simply name or rename the function event without extra passages.

**Comment by ivan-mogilko** [comment]
And how do you link same function to multiple events without manual linking (or annotations fwiw)?

**Comment by AlanDrake** [comment]
> And how do you link same function to multiple events without manual linking?

By using the editor. I just think defaults that are following the naming convention should sort themselves out without extra burden.

**Comment by ivan-mogilko** [comment]
If the delegates are implemented, the events become dynamic entities, meaning they may be linked and unlinked at runtime.
In my opinion, in such situation implicit automated linking by function name only becomes a contradiction to the delegate system, and a potential problem.

**Comment by messengerbag** [comment]
I think the biggest problem is that the process is fairly unintuitive for new users, and that even doing it doesn't guarantee that users form the right mental model/understanding of how it works (in part because clicking to create an event handler moves the focus over to the script). For example, one user today thought that clicking in the event pane was just a shortcut to generating the function header in the script, not realizing that it is auto-filling the field in the event pane that is the important part.

Improving efficiency/convenience is very much a secondary concern, IMO.

To me, that suggests two alternative paths: Changing the model of how this works completely, to something that is more intuitive. Or assisting users to avoid the mistake and understand what needs to be done.

Making event functions auto-link if they match the default name would be an example of the first approach, but IMO this has a number of drawbacks. A major one is that if you still retain the ability to link functions manually, the way it works becomes _less_ understandable, since now you have even more inconsistency.

Using warnings is an example of the second approach: stick with the way it works currently (for now), but help users to use it correctly. I think that's a better approach.

**Comment by messengerbag** [comment]
Would it be possible to have annotations as the in-script manifestation of the event property fields? IOW, whenever there is an event annotation to a function, the event field has the function name, and vice versa: the two are guaranteed to always be in sync.

If so, I can see that being fairly intuitive: the association is explicit in the script, so someone who copies an example can see it and include it. At the same time, we preserve the ability to view it from the other perspective in the event pane, and to auto-generate it with a click. (Will have to think about what the annotation should look like if the same function is linked with multiple events.)

This could even be extended to the global event handlers, so that they would no longer rely on "magic names." (That's a pretty major change, though, which needs careful consideration.)

**Comment by ericoporto** [comment]
Just a note that in Winforms, if you delete the code but keep the it in the properties menu VS designer crashes and this is still true on VS2022. They appear to be using the code to fill the properties but there's some illusion going there. Actual implementation of this in ags editor interface without code generation in a separate file (if the `event += func` was visible in script editor) would require a very good AGS script parser even with delegates.

The Kate Editor, which also uses Scintilla has linter support through LSP (#1067). When Atom was a thing I used a generic linter on top of the AGS Script treesitter grammar from edmundito and it picked up a few coding issues, so I guess there's some way to port something from looking at these two editors somehow - not sure how.

My perception is to either warn through compiler or perhaps better to do it in the editor directly through a linter - the obvious issue with the linter is that our Editor doesn't support/has one, so it isn't easy to "just add a new rule", and even in this case it would require some table of lateral information to be able to have it pick up if the editor has linked or not the function.

**Comment by ivan-mogilko** [comment]
There's another possible situation, which I witnessed recently, where user first created and linked these functions correctly, then renamed the Hotspot's name, and decided to also rename function names alongside, but did not change their names in the table.

I suppose that the Editor could check if function names match certain pattern, and if they do, then rename them automatically when the object's name changes.

**Comment by messengerbag** [comment]
If I understand you correctly, @ivan-mogilko, you're saying that when you change the name of entities in the editor (Characters, Hotspots, etc.), the IDE would check if there are any linked events for that entity, and if there are and their names contain the old entity name, automatically replace that substring with the new name? I guess that would be a nice accelerator feature, but I would probably ask the user for confirmation rather than make it completely automatic.

If it's not done and the user renames the function manually without changing the event property (or annotation, if those are introduced), there will be a broken reference in the event pane, and in that case I think the IDE should show a warning or error at build time. (Or in principle even as a "live" project status notification while editing; but that would probably require a completely different approach.) Maybe it already does so?

**Comment by ericoporto** [comment]
The warning messages mentioned were done in https://github.com/adventuregamestudio/ags/pull/2556 . I noticed double clicking on them doesn't show the pane that you would go to either delete the dangling event or jump from to a script (through the three dots property grid)

**Comment by ivan-mogilko** [comment]
> I noticed double clicking on them doesn't show the pane that you would go to either delete the dangling event or jump from to a script (through the three dots property grid)

Mmm... this would have to be researched. In theory that would require at least attaching an object reference to the message, but that likely should not be actual memory reference, but some key (type name, script name, etc). That's because we cannot predict whether the object will be still available in memory, or available at all in the project, when user clicks on the message.

**Comment by ericoporto** [comment]
I think only a string name of the object would be fine, there is a version of going to things pane that uses the type name that I used in the got to definition that was component based, but in theory all the data is too in the Project Explorer Tree, and going to it could be done by name as string in it.

**Comment by ivan-mogilko** [comment]
Right, objects must have unique name in the same namespace, so maybe type is not needed.
But it also will have to have Room index in case it's inside a room, and parent Index for GUI controls.

**Comment by ericoporto** [comment]
If there was a type of error message for missing handlers they could hold this information in them so these could be recovered from the entries in the output panel.

**Comment by ericoporto** [comment]
This issue was more focused on the problem itself, than in the solution, and that is fine, but I think #2603 and #2556 did things that should help reduce the problem. There are perhaps remaining issues (#1439), but perhaps this issue itself could be closed, as the Editor warns about events both ways (unlinked or linked but missing).
